probe_src/probe_frontend/build.rs

```diff
@@ -0,0 +1,118 @@
+        } else {
+            vec![]
+        }
+    }
+}
+
+fn main() {
+    // Tell cargo to look for shared libraries in the specified directory
+    // println!("cargo:rustc-link-search=/path/to/lib");
+
+    // Tell cargo to tell rustc to link the system bzip2
+    // shared library.
+    // println!("cargo:rustc-link-lib=bz2");
+
+    // The bindgen::Builder is the main entry point
+    // to bindgen, and lets you build up options for
+    // the resulting bindings.
+    let bindings = bindgen::Builder::default()
+        .header_contents(
+            "wrapper",
+            "
+            #define _GNU_SOURCE
+            #include <stdbool.h>
+            #include <stddef.h>
+            #include <stdint.h>
+            #include <sys/stat.h>
+            #include <sys/types.h>
+            #include <utime.h>
+
+            // HACK: defining this manually instead of using <sys/resource.h> is
+            // a huge hack, but it greatly reduces the generated code complexity
+            // since in glibc all the long ints are unions over two types that
+            // both alias to long int, this is done for kernel-userland
+            // compatibilityreasons that don't matter here.
+            struct rusage {
+                struct timeval ru_utime;
+                struct timeval ru_stime;
+                long int ru_maxrss;
+                long int ru_ixrss;
+                long int ru_idrss;
+                long int ru_isrss;
+                long int ru_minflt;
+                long int ru_majflt;
+                long int ru_nswap;
+                long int ru_inblock;
+                long int ru_oublock;
+                long int ru_msgsnd;
+                long int ru_msgrcv;
+                long int ru_nsignals;
+                long int ru_nvcsw;
+                long int ru_nivcsw;
+            };
+
+            #define BORROWED
+            #define OWNED
+        ",
+        )
+        // The input header we would like to generate
+        // bindings for.
+        .header("./include/prov_ops.h")
+        // .header_contents("sizeof", "
+        //     const size_t OP_SIZE = sizeof(struct Op);
+        // ")
+        // only parse the Op type (and any types contained within, recursively)
+        .allowlist_item("^(Op)$")
+        // Tell cargo to invalidate the built crate whenever any of the
+        // included header files changed.
+        .parse_callbacks(Box::new(bindgen::CargoCallbacks::new()))
+        .parse_callbacks(Box::new(SerdeDeriveCallback {}))
+        // Finish the builder and generate the bindings.
+        .generate()
+        // Unwrap the Result and panic on failure.
+        .expect("Unable to generate bindings");
+
+    // Write the bindings to the $OUT_DIR/bindings.rs file.
+    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
+    bindings
+        .write_to_file(out_path.join("bindings.rs"))
+        .expect("Couldn't write bindings!");
+}
```

Comment:
That seems reasonable, but I don't see the relevance to rust-bindgen. I think that would make sense as part of a future PR addressing my desire to refactor part of this code into a library. We could even go so far as to implement python bindings for such a library, since (last I checked) there are some decent tools for writing python libs in rust.

Reply:
Accomplished using proc-macros over the rust structs instead of by parsing `prov_ops.h` but still, done in https://github.com/charmoniumQ/PROBE/pull/20/commits/a83cce757de26a1bf559602fcd74f7f2b6df6705.